app/Actions/Album/Archive.php

```diff
@@ -141,12 +141,12 @@ private function compress_album($photos_sql, $dir_name, &$dirs, $parent_dir, $al
 		// cares in what order photos are zipped?
 
 		foreach ($photos as $photo) {
-			// For photos in public smart albums, skip the ones
-			// that are not downloadable based on their actual
-			// parent album.
+			// For photos in smart or tag albums, skip the ones that are not
+			// downloadable based on their actual parent album.  Photos from
+			// Unsorted are handled implicitly by the photo ownership test.
 			if (
-				$this->albumFactory->is_smart($albumID) && !AccessControl::is_logged_in() &&
-				$photo->album_id !== null && !$photo->album->is_downloadable()
+				$album->smart && !AccessControl::is_current_user($photo->owner_id) &&
+				!$photo->album->is_downloadable()
```

Comment:
You mean prefetching `$photo->album`? Yeah, I get that. But this is a condition that will rarely be checked so I don't think we should optimize for it.

I gave it some considerable thought last night and I convinced myself that `album_id` will never be `null` by the time the third case is reached. The reason is that in case of photos with `album_id = null`, the smart albums only show the ones belonging to the current user (or none in public mode). So any such photos will be caught by the second case (`AccessControl::is_current_user($photo->owner_id)`) and the third one will never be reached. This is captured in the comments (maybe in a little cryptic fashion :wink:).

Reply:
Having thought about it some more, I decided to handle the case of `album_id == null`. I don't think it can _currently_ happen, but why make the code needlessly fragile and subject to breaking after some seemingly unrelated future changes?